core/payments/client_ledger.go

```diff
@@ -0,0 +1,270 @@
+package payments
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/Layr-Labs/eigenda/api/clients/v2/metrics"
+	"github.com/Layr-Labs/eigenda/common/enforce"
+	"github.com/Layr-Labs/eigenda/core"
+	"github.com/Layr-Labs/eigenda/core/payments/ondemand"
+	"github.com/Layr-Labs/eigenda/core/payments/reservation"
+	"github.com/Layr-Labs/eigenda/encoding"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	gethcommon "github.com/ethereum/go-ethereum/common"
+)
+
+// The ClientLedger manages payment state for a single account. It is only used by *clients*, not by the disperser
+// or validator nodes.
+//
+// The ClientLedger aggressively triggers panics for errors that indicate no future payments will succeed. A client
+// is only useful if it can disperse blobs, and blobs can only be dispersed with a functioning payment mechanism.
+//
+// TODO(litt3): Currently, the client ledger has no mechanism to observe the following changes that may occur in the
+// PaymentVault:
+//
+// 1. Reservation expiration
+// 2. Reservation update
+// 3. OnDemand deposit
+//
+// It is the responsibility of the user to restart the client if such a change occurs. A mechanism should be implemented
+// to remove this burden from the user.
+type ClientLedger struct {
+	logger             logging.Logger
+	accountantMetricer metrics.AccountantMetricer
+	accountID          gethcommon.Address
+
+	// Though it would theoretically be possible to infer mode of operation based on on-chain state, it's important
+	// that this be directly configurable by the user, to ensure that reality matches intention.
+	//
+	// Consider, for example, if a user intends to operate with a reservation covering the majority of dispersals,
+	// with an on-demand balance as a backup. If there is a configuration issue which prevents the reservation from
+	// being used, the client could mistakenly burn through all backup funds before becoming aware of the
+	// misconfiguration. In such cases, it's better to fail early, to bring the misconfiguration to the attention of the
+	// user as soon as possible.
+	clientLedgerMode ClientLedgerMode
```

Comment:
thinking through this. the main user of this client will be proxy. I don't think (?) we want to add a proxy flag to set this? So we prob want to programmatically create the reservationLedger and onDemanderLedger anyways, and then set this mode also programmatically. In such a case I'm not sure I see the value.

Agree in general we prob should keep this and its generally good, just trying to fetch for your thoughts here and think through this a bit more.

Reply:
Agree. Actually thinking through this a bit more I also came to the conclusion that adding a proxy flag could actually make sense. can think through whether we want it optional or required later. if optional default would be to use both reservation and on-demand?